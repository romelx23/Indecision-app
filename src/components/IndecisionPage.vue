<!-- eslint-disable no-useless-return -->
<script>
export default {
  data () {
    return {
      question: '',
      answer: null,
      image: '',
      isValidQuestion: false
    }
  },
  methods: {
    async getAnswer () {
      this.answer = 'Pensando...'
      const response = await fetch('https://yesno.wtf/api')
      const { answer, image } = await response.json()
      this.answer = answer
      this.image = image
    //   console.log(answer, image)
    }
  },
  watch: {
    question (value, oldValue) {
      try {
        this.isValidQuestion = false
        console.log({ value })
        //   console.log(value.includes('?'))
        if (!value.includes('?')) return
        //   TODO realizar la peticion
        this.getAnswer()
        this.isValidQuestion = true
      } catch (error) {
        console.log('IndecisionPage', error)
        // eslint-disable-next-line no-unused-expressions
        this.answer = 'No se pudo cargar del API'
        this.image = null
      }
    }
  }
}
</script>

<template>
  <img v-if="image" alt="Vue logo" :src="image" />
  <div class="bg-dark"></div>
  <div class="container">
    <div class="indecision-container">
      <input
      v-model="question"
      type="text"
      placeholder="Hazme una pregunta"
      />
      <p>Recuerda terminar con un sgino de interrogaci√≥n (?)</p>
    </div>
    <div class="respuesta">
      <h2 v-if="isValidQuestion">{{question}}</h2>
      <h1>{{
        answer==="yes"?
        "Si"
        :answer==="no"?
        "No"
        :answer==="maybe"?
        "Tal vez"
        :answer==="Pensando..."?
        "Pensando..."
        :"Pensando..."
        }}</h1>
    </div>
  </div>
</template>

<style scoped>
body {
  position: relative;
}

h1 {
  margin: 0;
  padding: 0;
}
h2{
    font-size: 1.5rem;
}
img {
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  position: absolute;
  z-index: 0;
}

p {
  font-size: 1.2rem;
}

.bg-dark {
  /* overlay */
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  position: absolute;
  z-index: 1;
  background: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  padding: 30px 0;
}

.container {
  position: relative;
  z-index: 1;
  width: 100%;
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.respuesta {
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

input {
  width: 90%;
  padding: 15px;
  font-size: 18px;
  border: 1px solid #ccc;
  border-radius: 4px;
  margin-bottom: 10px;
}
</style>
