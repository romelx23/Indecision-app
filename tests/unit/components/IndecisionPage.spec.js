import { shallowMount } from '@vue/test-utils'
import IndecisionPage from '@/components/IndecisionPage.vue'
// En esta secci贸n vimos los Spy, Mocks y Wrapper de Vue Test Utils
// Spy: Es una funci贸n que se mantiene pendiente de una funci贸n que se ejecuta en el c贸digo.
// Mock :Es como va lucir mi data
// vm : es la intancia de vue donde se ecuentran las propiedad y metodos de mi componente
// wrapper : es una instancia de vue que se encarga de renderizar el componente
// En este caso vamos a crear un wrapper de IndecisionPage y vamos a verificar que el componente se renderice correctamente
describe('Indecision Component', () => {
  let wrapper
  let clgSpy
  global.fetch = jest.fn(() => Promise.resolve({
    json: () => Promise.resolve({
      answer: 'yes',
      forced: false,
      image: 'https://yesno.wtf/assets/yes/2.gif'
    })
  }))
  beforeEach(() => {
    wrapper = shallowMount(IndecisionPage)

    clgSpy = jest.spyOn(console, 'log')
    jest.clearAllMocks()
  })
  test('debe hacer match con el snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('Escribir en el input no debe de dispara nada console.log()', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    const input = wrapper.find('input')
    await input.setValue('hola')
    expect(clgSpy).toHaveBeenCalledTimes(1)
    expect(getAnswerSpy).toHaveBeenCalledTimes(0)
  })
  test('Escribir el simbolo de interrogacion debe lanzar el fetch ', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    const input = wrapper.find('input')
    await input.setValue('hola?')
    // expect(clgSpy).toHaveBeenCalledTimes(1)
    expect(getAnswerSpy).toHaveBeenCalled()
  })
  test('Pruebas en getAnswer()', async () => {
    await wrapper.vm.getAnswer()
    const img = wrapper.find('img')
    expect(img.exists()).toBeTruthy()
    expect(wrapper.vm.image).toBe('https://yesno.wtf/assets/yes/2.gif')
    expect(wrapper.vm.answer).toBe('yes')

    // console.log(wrapper.vm.image)
    // console.log(wrapper.vm.answer)
  })
  test('Pruebas en getAnswer-Fallo en el API', async () => {
    fetch.mockImplementationOnce(() => Promise.reject(new Error('Error')))
    await wrapper.vm.getAnswer()
    // TODO fallo en la API
    const img = wrapper.find('img')
    expect(img.exists()).toBeFalsy()
    expect(wrapper.vm.answer).toBe('No se pudo cargar del API')
  })
})
