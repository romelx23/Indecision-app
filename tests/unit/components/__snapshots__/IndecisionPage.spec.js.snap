// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe hacer match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="container">
  <div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
    <p>Recuerda terminar con un sgino de interrogaci√≥n (?)</p>
  </div>
  <div class="respuesta">
    <!--v-if-->
    <h1>Pensando...</h1>
  </div>
</div>
`;
